test-suite/squidconf/tls-certs

```diff
@@ -0,0 +1,192 @@
+# case: 1a, 2b, 3a invalid
+# case: 1b, 2b, 3a invalid
+# case: 1c, 2b, 3a invalid
+# case: 1d, 2b, 3a invalid
+
+# case: 1a, 2a, 3b
+https_port 127.0.2.4:8443 tls-cert=./ca-mid-rsa.crt tls-key=./ca-mid-rsa.pkey \
+	tls-generate-host-certificates=on
+https_port 127.0.2.1:8443 tls-cert=./ca-root-rsa.crt tls-key=./ca-root-rsa.pkey \
+	tls-generate-host-certificates=on
+# case: 1b, 2a, 3b
+https_port 127.0.1.2:8443 tls-cert=./leaf-chain-rsa.pem tls-key=./leaf-rsa.pkey \
+	tls-generate-host-certificates=on
+# case: 1c, 2a, 3b invalid
+# case: 1a, 2b, 3b N/A (requires cafile=/capath= options)
+# case: 1b, 2b, 3b
+https_port 127.0.2.7:8443 tls-cert=./ca-chain-rsa.pem \
+	tls-generate-host-certificates=on
+# case: 1c, 2b, 3b
+https_port 127.0.2.5:8443 tls-cert=./ca-mid-rsa.pem \
+	tls-generate-host-certificates=on
+# case: 1d, 2b, 3b invalid
+https_port 127.0.2.2:8443 tls-cert=./ca-root-rsa.pem \
+	tls-generate-host-certificates=on
+
+# case: 1a, 2a, 3c invalid
+# case: 1b, 2a, 3c invalid
+# case: 1c, 2a, 3c invalid
+# case: 1d, 2a, 3c invalid
+# case: 1a, 2b, 3c invalid
+# case: 1b, 2b, 3c invalid
+# case: 1c, 2b, 3c invalid
+
+# case: 1a, 2a, 3d
+https_port 127.0.2.3:8443 tls-cert=./ca-root-rsa.crt tls-key=./ca-root-rsa.pkey \
+	tls-generate-host-certificates
+# case: 1b, 2a, 3d invalid
+# case: 1c, 2a, 3d
+https_port 127.0.2.6:8443 tls-cert=./ca-mid-rsa.crt tls-key=./ca-mid-rsa.pkey \
+	tls-generate-host-certificates
+# case: 1d, 2a, 3d invalid
+# case: 1a, 2b, 3d invalid
+# case: 1b, 2b, 3d
+https_port 127.0.2.7:8443 tls-cert=./ca-chain-rsa.pem \
+	tls-generate-host-certificates
+# case: 1c, 2b, 3d
+https_port 127.0.2.4:8443 tls-cert=./ca-mid-rsa.crt tls-key=./ca-mid-rsa.pkey \
+	tls-generate-host-certificates
+# case: 1d, 2b, 3d invalid
+
+#
+# additional Squid-4+ permutations [REQUIRES PR#140]
+#
+
+# case: 1a, 2a, 3e not supported
+# case: 1b, 2a, 3e invalid
+# case: 1c, 2a, 3e invalid
+# case: 1d, 2a, 3e invalid
+#https_port 127.0.3.1:8443 tls-generate-host-certificates=./ca-root-rsa.crt
+# case: 1a, 2b, 3e invalid
+# case: 1b, 2b, 3e invalid
+# case: 1c, 2b, 3e
+#https_port 127.0.2.7:8443 tls-cert=./ca-chain-rsa.pem \
+#	tls-generate-host-certificates=./ca-chain-rsa.pem
+# case: 1d, 2b, 3e
+#https_port 127.0.3.2:8443 tls-generate-host-certificates=./ca-chain-rsa.pem
+#https_port 127.0.3.1:8443 tls-generate-host-certificates=./ca-root-rsa.pem
+
+#
+# Common misconfigurations that should result in Squid configuration failures.
+# TODO: Enable and check each individual negative test case in isolation.
+#
+
+# - missing key:
+#	WARNING: 'HTTPS_port 127.0.4.1:8443' missing private key in './leaf-rsa.crt'
+#	ERROR: Failed to initialize server context with keys from ./leaf-rsa.crt: The request is invalid.
+#	FATAL: HTTPS_port 127.0.4.1:8443 initialization error
+#https_port 127.0.4.1:8443 tls-cert=./leaf-rsa.crt
+
+# - missing CA cert
+#	XXX: Squid ignores the error silently.
+# https_port 127.0.4.2:8443 tls-generate-host-certificates=off
```

Comment:
This configuration can be fixed by adding a leaf certificate/key, right? If so, we should not label it "missing CA".

Reply:
The point was that Squid did not detect this as a misconfiguration at all. It happily just started operating. That is gone in current code, master produces an error.